File: ui/src/lib/formula.ts
export type FormulaProgram = {
  bindings: FormulaBinding[];
  body: FormulaExpr;
};

export type FormulaBinding = {
  name: string;
  value: FormulaExpr;
};

export type FormulaExpr =
  | { type: "number"; value: number; raw: string }
  | { type: "variable"; name: string }
  | { type: "argv"; slot: number }
  | { type: "unary"; op: "+" | "-"; expr: FormulaExpr }
  | { type: "binary"; op: "+" | "-" | "*" | "/" | "^"; left: FormulaExpr; right: FormulaExpr }
  | { type: "call"; name: string; args: FormulaExpr[] };

export type FormulaAnalysis =
  | { ok: true; program: FormulaProgram; tex: string }
  | { ok: false; error: string };

export const FORMULA_SYNTAX_OVERVIEW = [
  "numbers: 1, 2.5, .75, 1e3",
  "operators: +, -, *, /, ^",
  "functions: sqrt(x), nrt(x, n), abs(x), sin(x), cos(x), tan(x), ln(x), log(x), exp(x), min(a, b), max(a, b)",
  "constants: pi, e",
  "argv inputs: $1, $2, $3, ...",
  "let bindings: let x = 2; y = x^3 in y + $1",
].join("\n");

export function analyzeFormula(source: string): FormulaAnalysis {
  try {
    const program = parseFormula(source);
    return { ok: true, program, tex: renderProgramToTex(program) };
  } catch (error) {
    return { ok: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export function parseFormula(source: string): FormulaProgram {
  return new Parser(source).parseProgram();
}

export function renderProgramToTex(program: FormulaProgram): string {
  const substitutions = new Map<string, FormulaExpr>();
  for (const binding of program.bindings) {
    substitutions.set(binding.name, substituteExpr(binding.value, substitutions));
  }
  return renderExprToTex(substituteExpr(program.body, substitutions), 0);
}

function substituteExpr(expr: FormulaExpr, substitutions: Map<string, FormulaExpr>): FormulaExpr {
  switch (expr.type) {
    case "number":
    case "argv":
      return expr;
    case "variable":
      return substitutions.get(expr.name) ?? expr;
    case "unary":
      return { ...expr, expr: substituteExpr(expr.expr, substitutions) };
    case "binary":
      return {
        ...expr,
        left: substituteExpr(expr.left, substitutions),
        right: substituteExpr(expr.right, substitutions),
      };
    case "call":
      return { ...expr, args: expr.args.map((arg) => substituteExpr(arg, substitutions)) };
  }
}

function renderExprToTex(expr: FormulaExpr, parentPrecedence: number): string {
  switch (expr.type) {
    case "number":
      return expr.raw;
    case "variable":
      return `\\mathrm{${escapeText(expr.name)}}`;
    case "argv":
      return `\\mathrm{\\$${expr.slot}}`;
    case "unary": {
      const value = `${expr.op}${renderExprToTex(expr.expr, 4)}`;
      return parentPrecedence > 4 ? `\\left(${value}\\right)` : value;
    }
    case "binary": {
      const precedence = binaryPrecedence(expr.op);
      const left = renderExprToTex(expr.left, precedence);
      const right = renderExprToTex(expr.right, expr.op === "^" ? precedence - 1 : precedence + (expr.op === "-" ? 1 : 0));
      const rendered =
        expr.op === "*"
          ? `${left} \\cdot ${right}`
          : expr.op === "/"
            ? `\\frac{${renderExprToTex(expr.left, 0)}}{${renderExprToTex(expr.right, 0)}}`
            : expr.op === "^"
              ? `${left}^{${right}}`
              : `${left} ${expr.op} ${right}`;
      return parentPrecedence > precedence ? `\\left(${rendered}\\right)` : rendered;
    }
    case "call": {
      if (expr.name === "sqrt" && expr.args.length === 1) {
        return `\\sqrt{${renderExprToTex(expr.args[0], 0)}}`;
      }
      if (expr.name === "nrt" && expr.args.length === 2) {
        return `\\sqrt[${renderExprToTex(expr.args[1], 0)}]{${renderExprToTex(expr.args[0], 0)}}`;
      }
      const args = expr.args.map((arg) => renderExprToTex(arg, 0)).join(", ");
      return `\\operatorname{${escapeText(expr.name)}}\\left(${args}\\right)`;
    }
  }
}

function binaryPrecedence(op: FormulaExpr[Extract<keyof FormulaExpr, never>] | string): number {
  switch (op) {
    case "+":
    case "-":
      return 1;
    case "*":
    case "/":
      return 2;
    case "^":
      return 3;
    default:
      return 0;
  }
}

function escapeText(value: string) {
  return value.replace(/([{}_#$%&])/g, "\\\\$1");
}

type Token =
  | { type: "number"; value: number; raw: string }
  | { type: "ident"; value: string }
  | { type: "argv"; value: number }
  | { type: "let" | "in" | "+" | "-" | "*" | "/" | "^" | "(" | ")" | "," | ";" | "=" | "eof" };

class Parser {
  private cursor = 0;
  private current: Token = { type: "eof" };

  constructor(private readonly source: string) {
    this.bump();
  }

  parseProgram(): FormulaProgram {
    const bindings: FormulaBinding[] = [];
    if (this.current.type === "let") {
      this.bump();
      while (true) {
        const name = this.expectIdent("expected binding name after `let`");
        this.expect("=", "expected `=` in let binding");
        bindings.push({ name, value: this.parseExpr() });
        if (this.peekType() === ";") {
          this.bump();
          continue;
        }
        this.expect("in", "expected `;` or `in` after let binding");
        break;
      }
    }
    const body = this.parseExpr();
    if (this.current.type !== "eof") {
      throw new Error("unexpected trailing input");
    }
    return { bindings, body };
  }

  private parseExpr(): FormulaExpr {
    return this.parseAddSub();
  }

  private parseAddSub(): FormulaExpr {
    let expr = this.parseMulDiv();
    while (this.current.type === "+" || this.current.type === "-") {
      const op = this.current.type;
      this.bump();
      expr = { type: "binary", op, left: expr, right: this.parseMulDiv() };
    }
    return expr;
  }

  private parseMulDiv(): FormulaExpr {
    let expr = this.parsePower();
    while (this.current.type === "*" || this.current.type === "/") {
      const op = this.current.type;
      this.bump();
      expr = { type: "binary", op, left: expr, right: this.parsePower() };
    }
    return expr;
  }

  private parsePower(): FormulaExpr {
    const expr = this.parseUnary();
    if (this.current.type === "^") {
      this.bump();
      return { type: "binary", op: "^", left: expr, right: this.parsePower() };
    }
    return expr;
  }

  private parseUnary(): FormulaExpr {
    if (this.current.type === "+" || this.current.type === "-") {
      const op = this.current.type;
      this.bump();
      return { type: "unary", op, expr: this.parseUnary() };
    }
    return this.parsePrimary();
  }

  private parsePrimary(): FormulaExpr {
    switch (this.current.type) {
      case "number": {
        const token = this.current;
        this.bump();
        return { type: "number", value: token.value, raw: token.raw };
      }
      case "argv": {
        const slot = this.current.value;
        this.bump();
        return { type: "argv", slot };
      }
      case "ident": {
        const name = this.current.value;
        this.bump();
        if (this.peekType() === "(") {
          this.bump();
          const args: FormulaExpr[] = [];
          if (this.peekType() !== ")") {
            while (true) {
              args.push(this.parseExpr());
              if (this.peekType() === ",") {
                this.bump();
                continue;
              }
              break;
            }
          }
          this.expect(")", "expected `)` to close call");
          return { type: "call", name, args };
        }
        return { type: "variable", name };
      }
      case "(": {
        this.bump();
        const expr = this.parseExpr();
        this.expect(")", "expected `)`");
        return expr;
      }
      case "let":
        throw new Error("let expressions must appear at the start of the formula");
      default:
        throw new Error("expected an expression");
    }
  }


  private peekType(): Token["type"] {
    return this.current.type;
  }

  private expect(expected: Token["type"], message: string) {
    if (this.current.type !== expected) {
      throw new Error(message);
    }
    this.bump();
  }

  private expectIdent(message: string) {
    if (this.current.type !== "ident") {
      throw new Error(message);
    }
    const value = this.current.value;
    this.bump();
    return value;
  }

  private bump() {
    this.skipWhitespace();
    if (this.cursor >= this.source.length) {
      this.current = { type: "eof" };
      return;
    }
    const char = this.source[this.cursor]!;
    switch (char) {
      case "+":
      case "-":
      case "*":
      case "/":
      case "^":
      case "(":
      case ")":
      case ",":
      case ";":
      case "=":
        this.cursor += 1;
        this.current = { type: char };
        return;
      case "$": {
        this.cursor += 1;
        const digits = this.consumeWhile((value) => /\d/.test(value));
        if (!digits) {
          throw new Error("expected digits after `$`");
        }
        this.current = { type: "argv", value: Number(digits) };
        return;
      }
      default:
        if (/\d|\./.test(char)) {
          this.current = this.lexNumber();
          return;
        }
        if (/[A-Za-z_]/.test(char)) {
          const ident = this.consumeWhile((value) => /[A-Za-z0-9_]/.test(value));
          this.current = ident === "let" || ident === "in" ? { type: ident } : { type: "ident", value: ident };
          return;
        }
        throw new Error(`unexpected character \`${char}\``);
    }
  }

  private lexNumber(): Token {
    const start = this.cursor;
    let sawDigit = false;
    let sawDot = false;
    while (this.cursor < this.source.length) {
      const char = this.source[this.cursor]!;
      if (/\d/.test(char)) {
        sawDigit = true;
        this.cursor += 1;
      } else if (char === "." && !sawDot) {
        sawDot = true;
        this.cursor += 1;
      } else {
        break;
      }
    }
    if (/[eE]/.test(this.source[this.cursor] ?? "")) {
      this.cursor += 1;
      if (/[+-]/.test(this.source[this.cursor] ?? "")) {
        this.cursor += 1;
      }
      const exp = this.consumeWhile((value) => /\d/.test(value));
      if (!exp) {
        throw new Error("invalid scientific notation");
      }
    }
    if (!sawDigit) {
      throw new Error("invalid number literal");
    }
    const raw = this.source.slice(start, this.cursor);
    return { type: "number", value: Number(raw), raw };
  }

  private consumeWhile(predicate: (value: string) => boolean) {
    const start = this.cursor;
    while (this.cursor < this.source.length && predicate(this.source[this.cursor]!)) {
      this.cursor += 1;
    }
    return this.source.slice(start, this.cursor);
  }

  private skipWhitespace() {
    while (this.cursor < this.source.length && /\s/.test(this.source[this.cursor]!)) {
      this.cursor += 1;
    }
  }
}

File: kernel/src/formula.rs
use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq)]
pub struct Program {
    pub bindings: Vec<Binding>,
    pub body: Expr,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Binding {
    pub name: String,
    pub value: Expr,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Number(f64),
    Variable(String),
    Argv(usize),
    Unary {
        op: UnaryOp,
        expr: Box<Expr>,
    },
    Binary {
        left: Box<Expr>,
        op: BinaryOp,
        right: Box<Expr>,
    },
    Call {
        name: String,
        args: Vec<Expr>,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum UnaryOp {
    Plus,
    Negate,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BinaryOp {
    Add,
    Subtract,
    Multiply,
    Divide,
    Power,
}

#[derive(Debug, thiserror::Error, PartialEq)]
pub enum FormulaError {
    #[error("{0}")]
    Message(String),
}

pub fn parse(source: &str) -> Result<Program, FormulaError> {
    Parser::new(source).parse_program()
}

pub fn evaluate(source: &str, argv: &[String]) -> Result<String, FormulaError> {
    let program = parse(source)?;
    let value = eval_program(&program, argv)?;
    Ok(format_number(value))
}

fn eval_program(program: &Program, argv: &[String]) -> Result<f64, FormulaError> {
    let mut env = HashMap::new();
    for binding in &program.bindings {
        let value = eval_expr(&binding.value, argv, &env)?;
        env.insert(binding.name.clone(), value);
    }
    eval_expr(&program.body, argv, &env)
}

fn eval_expr(
    expr: &Expr,
    argv: &[String],
    env: &HashMap<String, f64>,
) -> Result<f64, FormulaError> {
    match expr {
        Expr::Number(value) => Ok(*value),
        Expr::Variable(name) => env
            .get(name)
            .copied()
            .or_else(|| match name.as_str() {
                "pi" => Some(std::f64::consts::PI),
                "e" => Some(std::f64::consts::E),
                _ => None,
            })
            .ok_or_else(|| FormulaError::Message(format!("unknown variable `{name}`"))),
        Expr::Argv(slot) => {
            let raw = argv
                .get(slot.saturating_sub(1))
                .ok_or_else(|| FormulaError::Message(format!("missing argument ${slot}")))?;
            raw.trim()
                .parse::<f64>()
                .map_err(|_| FormulaError::Message(format!("argument ${slot} is not a number")))
                .and_then(ensure_finite)
        }
        Expr::Unary { op, expr } => {
            let value = eval_expr(expr, argv, env)?;
            match op {
                UnaryOp::Plus => Ok(value),
                UnaryOp::Negate => ensure_finite(-value),
            }
        }
        Expr::Binary { left, op, right } => {
            let left = eval_expr(left, argv, env)?;
            let right = eval_expr(right, argv, env)?;
            let value = match op {
                BinaryOp::Add => left + right,
                BinaryOp::Subtract => left - right,
                BinaryOp::Multiply => left * right,
                BinaryOp::Divide => {
                    if right == 0.0 {
                        return Err(FormulaError::Message("division by zero".to_string()));
                    }
                    left / right
                }
                BinaryOp::Power => left.powf(right),
            };
            ensure_finite(value)
        }
        Expr::Call { name, args } => eval_call(name, args, argv, env),
    }
}

fn eval_call(
    name: &str,
    args: &[Expr],
    argv: &[String],
    env: &HashMap<String, f64>,
) -> Result<f64, FormulaError> {
    let values = args
        .iter()
        .map(|expr| eval_expr(expr, argv, env))
        .collect::<Result<Vec<_>, _>>()?;
    let value = match (name, values.as_slice()) {
        ("sqrt", [value]) => {
            if *value < 0.0 {
                return Err(FormulaError::Message("sqrt() domain error".to_string()));
            }
            value.sqrt()
        }
        ("nrt", [value, degree]) => {
            if *degree == 0.0 {
                return Err(FormulaError::Message(
                    "nrt() requires a non-zero degree".to_string(),
                ));
            }
            value.powf(1.0 / degree)
        }
        ("abs", [value]) => value.abs(),
        ("sin", [value]) => value.sin(),
        ("cos", [value]) => value.cos(),
        ("tan", [value]) => value.tan(),
        ("ln", [value]) => {
            if *value <= 0.0 {
                return Err(FormulaError::Message("ln() domain error".to_string()));
            }
            value.ln()
        }
        ("log", [value]) => {
            if *value <= 0.0 {
                return Err(FormulaError::Message("log() domain error".to_string()));
            }
            value.log10()
        }
        ("exp", [value]) => value.exp(),
        ("min", [left, right]) => left.min(*right),
        ("max", [left, right]) => left.max(*right),
        _ => {
            return Err(FormulaError::Message(format!(
                "unsupported call `{name}` with {} arguments",
                values.len()
            )))
        }
    };
    ensure_finite(value)
}

fn ensure_finite(value: f64) -> Result<f64, FormulaError> {
    if value.is_finite() {
        Ok(value)
    } else {
        Err(FormulaError::Message(
            "expression produced a non-finite value".to_string(),
        ))
    }
}

fn format_number(value: f64) -> String {
    if value == 0.0 {
        return "0".to_string();
    }
    if value.fract() == 0.0 {
        return format!("{value:.0}");
    }
    let mut out = format!("{value:.12}");
    while out.contains('.') && out.ends_with('0') {
        out.pop();
    }
    if out.ends_with('.') {
        out.pop();
    }
    out
}

#[derive(Debug, Clone, PartialEq)]
enum TokenKind {
    Number(f64),
    Ident(String),
    Argv(usize),
    Let,
    In,
    Plus,
    Minus,
    Star,
    Slash,
    Caret,
    LParen,
    RParen,
    Comma,
    Semi,
    Equal,
    Eof,
}

struct Parser<'a> {
    source: &'a str,
    cursor: usize,
    current: TokenKind,
}

impl<'a> Parser<'a> {
    fn new(source: &'a str) -> Self {
        let mut parser = Self {
            source,
            cursor: 0,
            current: TokenKind::Eof,
        };
        parser.bump().expect("initial token");
        parser
    }

    fn parse_program(&mut self) -> Result<Program, FormulaError> {
        let mut bindings = Vec::new();
        if self.current == TokenKind::Let {
            self.bump()?;
            loop {
                let name = match &self.current {
                    TokenKind::Ident(name) => name.clone(),
                    _ => return self.error("expected binding name after `let`"),
                };
                self.bump()?;
                self.expect(TokenKind::Equal, "expected `=` in let binding")?;
                let value = self.parse_expr()?;
                bindings.push(Binding { name, value });
                match self.current {
                    TokenKind::Semi => {
                        self.bump()?;
                    }
                    TokenKind::In => {
                        self.bump()?;
                        break;
                    }
                    _ => return self.error("expected `;` or `in` after let binding"),
                }
            }
        }
        let body = self.parse_expr()?;
        if self.current != TokenKind::Eof {
            return self.error("unexpected trailing input");
        }
        Ok(Program { bindings, body })
    }

    fn parse_expr(&mut self) -> Result<Expr, FormulaError> {
        self.parse_add_sub()
    }

    fn parse_add_sub(&mut self) -> Result<Expr, FormulaError> {
        let mut expr = self.parse_mul_div()?;
        loop {
            let op = match self.current {
                TokenKind::Plus => BinaryOp::Add,
                TokenKind::Minus => BinaryOp::Subtract,
                _ => break,
            };
            self.bump()?;
            let right = self.parse_mul_div()?;
            expr = Expr::Binary {
                left: Box::new(expr),
                op,
                right: Box::new(right),
            };
        }
        Ok(expr)
    }

    fn parse_mul_div(&mut self) -> Result<Expr, FormulaError> {
        let mut expr = self.parse_power()?;
        loop {
            let op = match self.current {
                TokenKind::Star => BinaryOp::Multiply,
                TokenKind::Slash => BinaryOp::Divide,
                _ => break,
            };
            self.bump()?;
            let right = self.parse_power()?;
            expr = Expr::Binary {
                left: Box::new(expr),
                op,
                right: Box::new(right),
            };
        }
        Ok(expr)
    }

    fn parse_power(&mut self) -> Result<Expr, FormulaError> {
        let expr = self.parse_unary()?;
        if self.current == TokenKind::Caret {
            self.bump()?;
            let right = self.parse_power()?;
            Ok(Expr::Binary {
                left: Box::new(expr),
                op: BinaryOp::Power,
                right: Box::new(right),
            })
        } else {
            Ok(expr)
        }
    }

    fn parse_unary(&mut self) -> Result<Expr, FormulaError> {
        match self.current {
            TokenKind::Plus => {
                self.bump()?;
                Ok(Expr::Unary {
                    op: UnaryOp::Plus,
                    expr: Box::new(self.parse_unary()?),
                })
            }
            TokenKind::Minus => {
                self.bump()?;
                Ok(Expr::Unary {
                    op: UnaryOp::Negate,
                    expr: Box::new(self.parse_unary()?),
                })
            }
            _ => self.parse_primary(),
        }
    }

    fn parse_primary(&mut self) -> Result<Expr, FormulaError> {
        match &self.current {
            TokenKind::Number(value) => {
                let expr = Expr::Number(*value);
                self.bump()?;
                Ok(expr)
            }
            TokenKind::Argv(slot) => {
                let expr = Expr::Argv(*slot);
                self.bump()?;
                Ok(expr)
            }
            TokenKind::Ident(name) => {
                let name = name.clone();
                self.bump()?;
                if self.current == TokenKind::LParen {
                    self.bump()?;
                    let mut args = Vec::new();
                    if self.current != TokenKind::RParen {
                        loop {
                            args.push(self.parse_expr()?);
                            if self.current == TokenKind::Comma {
                                self.bump()?;
                                continue;
                            }
                            break;
                        }
                    }
                    self.expect(TokenKind::RParen, "expected `)` to close call")?;
                    Ok(Expr::Call { name, args })
                } else {
                    Ok(Expr::Variable(name))
                }
            }
            TokenKind::LParen => {
                self.bump()?;
                let expr = self.parse_expr()?;
                self.expect(TokenKind::RParen, "expected `)`")?;
                Ok(expr)
            }
            TokenKind::Let => self.error("let expressions must appear at the start of the formula"),
            _ => self.error("expected an expression"),
        }
    }

    fn expect(&mut self, expected: TokenKind, message: &str) -> Result<(), FormulaError> {
        if self.current == expected {
            self.bump()?;
            Ok(())
        } else {
            self.error(message)
        }
    }

    fn bump(&mut self) -> Result<(), FormulaError> {
        self.skip_ws();
        if self.cursor >= self.source.len() {
            self.current = TokenKind::Eof;
            return Ok(());
        }
        let rest = &self.source[self.cursor..];
        let mut chars = rest.char_indices();
        let (_, ch) = chars.next().expect("char");
        self.current = match ch {
            '+' => {
                self.cursor += 1;
                TokenKind::Plus
            }
            '-' => {
                self.cursor += 1;
                TokenKind::Minus
            }
            '*' => {
                self.cursor += 1;
                TokenKind::Star
            }
            '/' => {
                self.cursor += 1;
                TokenKind::Slash
            }
            '^' => {
                self.cursor += 1;
                TokenKind::Caret
            }
            '(' => {
                self.cursor += 1;
                TokenKind::LParen
            }
            ')' => {
                self.cursor += 1;
                TokenKind::RParen
            }
            ',' => {
                self.cursor += 1;
                TokenKind::Comma
            }
            ';' => {
                self.cursor += 1;
                TokenKind::Semi
            }
            '=' => {
                self.cursor += 1;
                TokenKind::Equal
            }
            '$' => {
                self.cursor += 1;
                let digits = self.consume_while(|c| c.is_ascii_digit());
                if digits.is_empty() {
                    return self.error("expected digits after `$`");
                }
                let slot = digits
                    .parse::<usize>()
                    .map_err(|_| FormulaError::Message("invalid argv slot".to_string()))?;
                TokenKind::Argv(slot)
            }
            c if c.is_ascii_digit() || c == '.' => self.lex_number()?,
            c if c.is_ascii_alphabetic() || c == '_' => self.lex_ident(),
            _ => return self.error(&format!("unexpected character `{ch}`")),
        };
        Ok(())
    }

    fn lex_number(&mut self) -> Result<TokenKind, FormulaError> {
        let start = self.cursor;
        let mut saw_digit = false;
        let mut saw_dot = false;
        while let Some(ch) = self.peek_char() {
            if ch.is_ascii_digit() {
                saw_digit = true;
                self.cursor += ch.len_utf8();
            } else if ch == '.' && !saw_dot {
                saw_dot = true;
                self.cursor += 1;
            } else {
                break;
            }
        }
        if matches!(self.peek_char(), Some('e' | 'E')) {
            self.cursor += 1;
            if matches!(self.peek_char(), Some('+' | '-')) {
                self.cursor += 1;
            }
            let exp = self.consume_while(|c| c.is_ascii_digit());
            if exp.is_empty() {
                return self.error("invalid scientific notation");
            }
        }
        if !saw_digit {
            return self.error("invalid number literal");
        }
        let value = self.source[start..self.cursor]
            .parse::<f64>()
            .map_err(|_| FormulaError::Message("invalid number literal".to_string()))?;
        Ok(TokenKind::Number(value))
    }

    fn lex_ident(&mut self) -> TokenKind {
        let ident = self.consume_while(|c| c.is_ascii_alphanumeric() || c == '_');
        match ident.as_str() {
            "let" => TokenKind::Let,
            "in" => TokenKind::In,
            _ => TokenKind::Ident(ident),
        }
    }

    fn consume_while(&mut self, mut predicate: impl FnMut(char) -> bool) -> String {
        let start = self.cursor;
        while let Some(ch) = self.peek_char() {
            if predicate(ch) {
                self.cursor += ch.len_utf8();
            } else {
                break;
            }
        }
        self.source[start..self.cursor].to_string()
    }

    fn peek_char(&self) -> Option<char> {
        self.source[self.cursor..].chars().next()
    }

    fn skip_ws(&mut self) {
        while let Some(ch) = self.peek_char() {
            if ch.is_whitespace() {
                self.cursor += ch.len_utf8();
            } else {
                break;
            }
        }
    }

    fn error<T>(&self, message: &str) -> Result<T, FormulaError> {
        Err(FormulaError::Message(message.to_string()))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_let_bindings() {
        let program = parse("let x = 2; y = x^3 in y + 1").expect("parse");
        assert_eq!(program.bindings.len(), 2);
    }

    #[test]
    fn evaluates_arithmetic_and_functions() {
        assert_eq!(evaluate("sqrt(16) + nrt(27, 3)", &[]).expect("eval"), "7");
        assert_eq!(evaluate("2^3^2", &[]).expect("eval"), "512");
        assert_eq!(evaluate("let x = 1.5 in x * 2", &[]).expect("eval"), "3");
    }

    #[test]
    fn evaluates_argv_inputs() {
        assert_eq!(
            evaluate("$1 + $2", &["4".to_string(), "5.5".to_string()]).expect("eval"),
            "9.5"
        );
    }

    #[test]
    fn reports_invalid_argv() {
        let error = evaluate("$2", &["1".to_string()]).expect_err("missing argv should fail");
        assert_eq!(
            error,
            FormulaError::Message("missing argument $2".to_string())
        );
    }
}
